### PR TITLE
Integration of pygrb aligntotalspin functionality into pycbc/transforms.py as a class

### PR DESCRIPTION
**Pull request by SamuelH-97**
Proposing adding the aligntotalspin function from pygrb to the transforms.py code as a callable class. This does require the import of numpy and import of the jframe_to_l0frame function from pycbc.waveform.waveform_modes currently.

The class is able to rigidly rotate a binary so that the total angular momentum has the given inclination (iota) instead of the orbital angular momentum. It returns the new inclination, s1, and s2. s1 and s2 are dimensionless spin. The spins are assumed to be given in the frame defined by the orbital angular momentum.

The class has been written to fit with the style of other classes in the transforms.py code.

This is primarily used in the generation of injected binaries used to test the sensitivity of targeted GRB searches.

### CODE REVIEW

**Comment by SamuelH-97** [comment]
I should add this is my first time doing a pull request or contributing here so please feel free to leave comments on if I did anything wrong with my request too

**Comment by spxiwh** [comment]
@SamuelH-97 Thanks for providing a first PR! It looks like you've set up everything correctly, and I've just approved the test code to run.

I'll ask @cdcapano if he would be willing to look over the content here, as this is something he knows much better than me. (@cdcapano if you can't do this, can you suggest someone else to look this over).

There are some style issues here that should be addressed first though. We do ask that everyone follows PEP-8 (for code) and the numpydoc (for documentation) styles. The automated code checker has picked up on a number of issues that should be addressed:

https://codeclimate.com/github/gwastro/pycbc/pull/3756

In most cases these should be trivial to fix, but in some cases (e.g. code duplication) it *might* make sense to keep what you have, but you should comment on that once you've resolved everything that you can.

I'll also leave a few specific comments on code style stuff in-line.

File: pycbc/transforms.py [hunk 1]
```diff
@@ -612,7 +612,101 @@ def inverse_jacobian(self, maps):
         ref_mass=1.4
         mchirp = maps['mchirp']
         return (2.**(-1./5) * self.ref_mass / mchirp)**(5./6)
+    
+######################################################################################################### TEST ################################################################################################################################################
```
Comment (spxiwh) [inline pycbc/transforms.py:616]: Remove comments like this before this get merged.

File: pycbc/transforms.py [hunk 2]
```diff
@@ -612,7 +612,101 @@ def inverse_jacobian(self, maps):
...
+import numpy as np
```
Comment (spxiwh) [inline pycbc/transforms.py:617]: This code already imports numpy, don't need to do it again.

File: pycbc/transforms.py [hunk 3]
```diff
@@ -612,7 +612,101 @@ def inverse_jacobian(self, maps):
...
+from pycbc.waveform.waveform_modes import jframe_to_l0frame
```
Comment (spxiwh) [inline pycbc/transforms.py:618]: Just to check (and not part of this PR). Have you checked that this code actually works? I'm not sure if it's been used for anything important ...

File: pycbc/transforms.py [hunk 4]
```diff
@@ -612,7 +612,101 @@ def inverse_jacobian(self, maps):
...
+
+class alignTotalSpin(BaseTransform):
+
```
Comment (spxiwh) [inline pycbc/transforms.py:621]: Remove whitespace here

File: pycbc/transforms.py [hunk 5]
```diff
@@ -612,7 +612,101 @@ def inverse_jacobian(self, maps):
...
         return (2.**(-1./5) * self.ref_mass / mchirp)**(5./6)
+    
+######################################################################################################### TEST ################################################################################################################################################
+import numpy as np
+from pycbc.waveform.waveform_modes import jframe_to_l0frame
+
+class alignTotalSpin(BaseTransform):
+    
+    """Parameters
+    ----------
```
Comment (spxiwh) [inline pycbc/transforms.py:623]: This isn't following the numpydoc formatting, and won't display correctly in the online docs. Can you update this, using other code as an example?

File: pycbc/transforms.py [hunk 6]
```diff
@@ -612,7 +612,101 @@ def inverse_jacobian(self, maps):
...
+        self.m1 = m1
+        self.m2 = m2
+        self.f_lower = f_lower
+        self.phi_ref = phi_ref
+        self.newIota = newIota
+        self.news1 = news1
+        self.news2 = news2
+        self._inputs = [self.iota, self.s1, self.s2, self.m1, self.m2, self.f_lower, self.phi_ref]
+        self._outputs = [self.newIota, self.news1, self.news2]
+        super(alignTotalSpin, self).__init__()   ############################################## what??????
```
Comment (spxiwh) [inline pycbc/transforms.py:653]: Why the comment here? (While we are still using python2 we do need to support old-style super statements .... python2 will be retired soon though, so feel free to update once we do that!)

File: pycbc/transforms.py [hunk 7]
```diff
@@ -612,7 +612,101 @@ def inverse_jacobian(self, maps):
...
+        phi_ref = self.phi_ref
+
+        # Calculate the quantities required by jframe_to_l0frame
+        if np.linalg.norm(maps[s1][:2]) != 0 or np.linalg.norm(maps[s1][:2]) != 0:
+            thetaJN = maps[iota]
+            phiJL = np.pi
+            l = np.array([0,0,1.])
+            theta1 = np.arccos(np.dot(maps[s1], l)/np.linalg.norm(maps[s1]))
+            theta2 = np.arccos(np.dot(maps[s2], l)/np.linalg.norm(maps[s2]))
+            phi12 = np.arccos(np.dot(maps[s1][:2], maps[s2][:2]) / np.linalg.norm(maps[s1][:2]) / \
```
Comment (spxiwh) [inline pycbc/transforms.py:679]: In general multi-line statements like this are hard to parse. PEP-8 does have rules on this (and CodeClimate will pick up on this). I would probably align on the brackets, and the newline symbol is not needed.

File: pycbc/transforms.py [hunk 8]
```diff
@@ -612,7 +612,101 @@ def inverse_jacobian(self, maps):
...
+                              np.linalg.norm(maps[s2][:2]))
+            
+            #print(maps[s1])
```
Comment (spxiwh) [inline pycbc/transforms.py:682]: Remove debugging statements before this gets onto the main branch

File: pycbc/transforms.py [hunk 9]
```diff
@@ -612,7 +612,101 @@ def inverse_jacobian(self, maps):
...
+
+            news1 = np.zeros(3)
+            news2 = np.zeros(3)
+            
+            output = \
```
Comment (spxiwh) [inline pycbc/transforms.py:687]: There is a tool called `black` that you can use to align small blocks of code like this to be PEP-8 compliant .... I know this stuff sounds picky, but uniform code is much easier to parse in a long module like this one.

File: pycbc/transforms.py [hunk 10]
```diff
@@ -612,7 +612,101 @@ def inverse_jacobian(self, maps):
...
+            jframe_to_l0frame(maps[m1], maps[m2], maps[f_lower], maps[phi_ref], thetaJN, phiJL,
+                      np.linalg.norm(maps[s1]), np.linalg.norm(maps[s2]),
+                      theta1, theta2,
+                      phi12)
+            
+            newIota = output['inclination']      #not elegant, just quickly done for testing
```
Comment (spxiwh) [inline pycbc/transforms.py:693]: If this is a "FIXME", then put it on the line above and use `# FIXME` (or fix it).

File: pycbc/transforms.py [hunk 11]
```diff
@@ -612,7 +612,101 @@ def inverse_jacobian(self, maps):
...
+            news1[0] = output['spin1x']
+            news1[1] = output['spin1y']
+            news1[2] = output['spin1z']
+            news2[0] = output['spin2x']
+            news2[1] = output['spin2y']
+            news2[2] = output['spin2z']
+            
+            #print(output['spin1x'])   ### WIP!!!!
```
Comment (spxiwh) [inline pycbc/transforms.py:701]: Remove

File: pycbc/transforms.py [hunk 12]
```diff
@@ -612,7 +612,101 @@ def inverse_jacobian(self, maps):
...
+            
+            out = {self.newIota: newIota, self.news1: news1, self.news2: news2}
+            return self.format_output(maps, out)
+        else:
+            out = {self.newIota: iota, self.news1: s1, self.news2: s2}
+            return self.format_output(maps, out)
+        
+######################################################################################################### TEST ################################################################################################################################################
```
Comment (spxiwh) [inline pycbc/transforms.py:709]: Remove this as well

**Review by spxiwh — CHANGES_REQUESTED** [review]

File: pycbc/transforms.py (hunk 12 above)
Comment (SamuelH-97) [reply pycbc/transforms.py:709]: Hi, thank you for the feedback, and thanks for making me aware of issues I should have resolved before doing the PR. I'll take some time today to have a look over these and hopefully get this in a much better state.

File: pycbc/transforms.py [hunk 13]
```diff
@@ -28,10 +28,12 @@
 from pycbc.waveform import parameters
 from pycbc.boundaries import Bounds
 from pycbc import VARARGS_DELIM
+from pycbc.waveform.waveform_modes import jframe_to_l0frame
```
Comment (cdcapano) [inline pycbc/transforms.py:31]: This might cause circular imports. It's usually a bad idea to import functions from sub modules in top-level modules. I originally put this function in `waveform_modes` because that's where it was being used, but it would be useful at a higher level. Maybe move it to `pycbc/pnutils.py`? Then import from there (also make sure to add an import to `pycbc/waveform/waveform_modes.py` too).

**Review by cdcapano — CHANGES_REQUESTED** [review]
This looks fine to me, although I agree with @spxiwh that you should address the PEP-8 and documentation issues (which looks like you're working on). My only additional request is you move the `jframe_to_l0frame` function to avoid circular imports; see my comment below.

File: pycbc/transforms.py (hunk 13 above)
Comment (SamuelH-97) [reply pycbc/transforms.py:31]: Thanks for the help, I'll look into this asap

**Comment by SamuelH-97** [comment]
I used black as suggested by @spxiwh to resolve many of the PEP-8 issues. The only flags I'm seeing in the codeclimate page now are for similar lines of code (which I think need to be that way), and lines that are too long. Are these still things you'd suggest I need to fix? Should I just break the long lines?

**Comment by spxiwh** [comment]
@SamuelH-97 Thanks for this. In some cases here I think `black` should have been able to keep the 79 character line limit. I think you probably ran it without specifying this. I quote from the `black` documentation:
```
If you’re paid by the line of code you write, you can pass --line-length with a lower number. Black will try to respect that. However, sometimes it won’t be able to without breaking other rules. In those rare cases, auto-formatted code will exceed your allotted limit.
```
In some cases here it probably will keep > 79 character lines if you use `line-length`, but please do try running `black` with `line-length 79` first. Then I think this is good to go.

I agree that the code repetition point here is not a problem. It would add confusion and reduce clarity to make this a function called twice.

File: pycbc/transforms.py (hunk 13 above)
Comment (spxiwh) [reply pycbc/transforms.py:31]: I would agree with Collin that if this function becomes useful outside of `waveform_modes` then it belongs somewhere else. Collin's moved this stuff about most recently, so his suggestion seems sensible.

File: pycbc/transforms.py [hunk 14]
```diff
@@ -592,26 +596,133 @@ def inverse_transform(self, maps):
...
+
+        iota = self.iota
+        s1 = self.s1
+        s2 = self.s2
+        m1 = self.m1
+        m2 = self.m2
+        f_lower = self.f_lower
+        phi_ref = self.phi_ref
+
+        if numpy.linalg.norm(maps[s1][:2]) != 0 or numpy.linalg.norm(maps[s1][:2]) != 0:
```
Comment (a-r-williamson) [inline pycbc/transforms.py:683]: This can be replaced without calling the numpy function
```suggestion
        if not all(s == 0.0 for s in s1[:2]) or not all(s == 0.0 for s in s2[:2]):
```

File: pycbc/transforms.py [hunk 15]
```diff
@@ -592,51 +596,50 @@ def inverse_transform(self, maps):
...
+        The name of the f_lower parameter
+    phi_ref : str
+        The name of the phi_ref parameter
     """
-    name = "Align_total_spin"
-    
+
+    name = "align_total_spin"
+
     def __init__(self, iota, s1, s2, m1, m2, f_lower, phi_ref, newIota, news1, news2):
```
Comment (a-r-williamson) [inline pycbc/transforms.py:643]: I think we should be passing in & out the x/y/z components of the spin, rather than packaging them into vectors.

**Comment by a-r-williamson** [comment]
The changes up until now have been tested and work if the suggested bugfix in #3760 is included.

**Comment by SamuelH-97** [comment]
Hi, sorry for the extended break on this request. I've moved the jframe_to_i0frame to pnutils.py as suggested. I assume I should probably do the same with it's sibling i0frame_to_jframe.

**Comment by SamuelH-97** [comment]
With these changes and fixes now pushed, would it be possible for you to take another look @cdcapano @spxiwh and make sure you're happy with them?

**Comment by ahnitz** [comment]
@a-r-williamson @SamuelH-97 Can you rebase so we can verify the unittests / CC pass?

**Comment by ahnitz** [comment]
@a-r-williamson @SamuelH-97 Are you sure this branch is as you've tested? The unittests are all failing for a sensible reason. The _formatdocstr function was private to the module you move the function from and hasn't been imported over.

**Comment by SamuelH-97** [comment]
> @a-r-williamson @SamuelH-97 Can you rebase so we can verify the unittests / CC pass?

Rebase is done

**Comment by SamuelH-97** [comment]
> @a-r-williamson @SamuelH-97 Are you sure this branch is as you've tested? The unittests are all failing for a sensible reason. The _formatdocstr function was private to the module you move the function from and hasn't been imported over.

Will get that fixed asap

File: pycbc/pnutils.py [hunk 16]
```diff
@@ -956,3 +956,108 @@ def hybrid_meco_frequency(m1, m2, chi1, chi2, qm1=None, qm2=None):
...
+                      spin1_a=0., spin2_a=0.,
+                      spin1_polar=0., spin2_polar=0.,
+                      spin12_deltaphi=0.):
+    """Converts J-frame parameters into L0 frame.
+
+    Parameters
+    ----------
+    {mass1}
+    {mass2}
+    {f_ref}
```
Comment (a-r-williamson) [inline pycbc/pnutils.py:971]: These and similar parts of the docstrings (parameters inside {} in inputs and outputs) will need to be written out explicitly as they're not going to be parsed like they would be in `waveform_modes.py`. For example:
```suggestion
    mass1 : float
        First component mass in solar mass units
    mass2 : float
        Second component mass in solar mass units
    f_ref: float
        Reference frequency
```

**Comment by titodalcanton** [comment]
Please rebase on master again, as we recently did some documentation fixes that may affect this. Sorry!

**Comment by titodalcanton** [comment]
@SamuelH-97 this is now breaking the documentation build and I think I know why: please have a look at the current docstring of `l0frame_to_jframe` and see how it was recently fixed (#3879). You need to carry over that fix here.

**Comment by a-r-williamson** [comment]
@ahnitz Hi Alex, @SamuelH-97 and I are a little stumped by the error in the 3.9 doc build test
https://github.com/gwastro/pycbc/runs/4457218789?check_suite_focus=true

```
Warning, treated as error:
/home/runner/work/pycbc/pycbc/.tox/py-docs/lib/python3.9/site-packages/pycbc/pnutils.py:docstring of pycbc.pnutils.jframe_to_l0frame:29:Unexpected indentation.
make: *** [Makefile:51: html] Error 2
```

If I can count, it looks like it doesn't like this line?
https://github.com/gwastro/pycbc/blob/89c39c522c4b10c88895816366ed008cfd58f8b2/pycbc/pnutils.py#L993

Do you know why?

**Comment by titodalcanton** [comment]
@a-r-williamson try adding a blank line between "Dictionary of:" and the first bullet.

**Comment by SamuelH-97** [comment]
@titodalcanton @a-r-williamson after adding the suggested blank lines the code tests seem to keep failing. Although on a quick glance this seems to be some failure in the tests themselves rather than my code. I could be wrong here, how does it look to you?

**Comment by titodalcanton** [comment]
@SamuelH-97 we are having problems at the moment due to new releases of python-ligo-lw and lalsuite, which break backwards compatibility. See #3901 for the fixes; you should rebase off master when that is merged.

**Comment by SamuelH-97** [comment]
> @SamuelH-97 we are having problems at the moment due to new releases of python-ligo-lw and lalsuite, which break backwards compatibility. See #3901 for the fixes; you should rebase off master when that is merged.

Oh I see, will have a look

**Comment by a-r-williamson** [comment]
The previous check failures are now resolved.

**Comment by a-r-williamson** [comment]
@ahnitz Given the current state of this, can it be merged? It's been tested and the builds are passing.

**Review by spxiwh — APPROVED** [review]
